File: u3dmobile/Assets/Game/FairyGUI/UIComDirector.cs
//use the menu item "U3DMOBILE/Install FairyGUI Runtime" to install the runtime,
//and add "U3DMOBILE_USE_FAIRYGUI" on the project setting "Scripting Define Symbols".
#if U3DMOBILE_USE_FAIRYGUI

using FairyGUI;

namespace U3DMobile
{
    class UIComDirector
    {
        private bool          _filled;
        private GComponent    _theGCom;
        private PackageSource _pkgSource;
        private string        _pkgName;
        private string        _comName;

        private Window _window ;
        private UICom  _com    ;
        private bool   _created;
        private bool   _shown  ;

        public Window window { get { return _window; } }
        public UICom  com    { get { return _com   ; } }
        public bool   shown  { get { return _shown ; } }

        public void SetGCom(GComponent aGCom)
        {
            if (_filled)
            {
                Log.Error($"can not set ui com repeatedly");
                return;
            }

            if (aGCom != null)
            {
                _filled  = true ;
                _theGCom = aGCom;
            }
        }

        public void SetResource(PackageSource source, string pkgName, string comName)
        {
            if (_filled)
            {
                Log.Error($"can not set package resource repeatedly");
                return;
            }

            if (string.IsNullOrWhiteSpace(pkgName))
            {
                Log.Error($"try to set empty package name");
                return;
            }
            if (string.IsNullOrWhiteSpace(comName))
            {
                Log.Error($"try to set empty component name");
                return;
            }

            _filled    = true   ;
            _pkgSource = source ;
            _pkgName   = pkgName;
            _comName   = comName;
        }

        public void Show()
        {
            if (!_filled)
            {
                return;
            }

            if (!_created)
            {
                //actions:
                if (_theGCom == null)
                {
                    PackageManager manager = PackageManager.instance;
                    GObject aGObject = manager.CreateGObject(_pkgSource, _pkgName, _comName);

                    _theGCom = aGObject.asCom;
                    if (_theGCom == null)
                    {
                        Log.Error($"failed to create '{_pkgName}/{_comName}'");
                        return;
                    }
                }

                _window = new Window
                {
                    contentPane = _theGCom,
                };

                _com = new UICom();
                _com.SetGCom(_theGCom, _comName);
                _com.Bind(this);

                //notification.
                _created = true;
                OnCreate();
            }
            if (!_shown)
            {
                //notification.
                _shown = true;
                OnShow();

                //actions.
                _window.Show();
            }
        }

        public void Hide()
        {
            if (_shown)
            {
                //actions.
                _window.Hide();
                
                //notification.
                _shown = false;
                OnHide();
            }
        }

        public void Dispose()
        {
            if (_shown)
            {
                //notification.
                _shown = false;
                OnHide();
            }
            if (_created)
            {
                //actions:
                _com = null;

                _window.Dispose();
                _window = null;

                _theGCom = null;

                //notification.
                _created = false;
                OnDestroy();
            }
        }

        protected virtual void OnCreate () {}
        protected virtual void OnShow   () {}
        protected virtual void OnHide   () {}
        protected virtual void OnDestroy() {}
    }
}

#endif

File: tsproj/manuscript/basics/UIComDirector.ts
import { FairyGUI  } from 'csharp'
import { Log       } from './Log'
import { U3DMobile } from 'csharp'
import { UICom     } from './UICom'

export class UIComDirector {

    private _filled   : boolean
    private _theGCom  : FairyGUI.GComponent
    private _pkgSource: U3DMobile.PackageSource
    private _pkgName  : string
    private _comName  : string

    private _window : FairyGUI.Window
    private _com    : UICom
    private _created: boolean
    private _shown  : boolean

    public get window(): FairyGUI.Window { return this._window }
    public get com   (): UICom           { return this._com    }
    public get shown (): boolean         { return this._shown  }

    public SetGCom(aGCom: FairyGUI.GComponent): void {
        if (this._filled) {
            Log.Error(`can not set ui com repeatedly`)
            return
        }

        if (aGCom) {
            this._filled  = true
            this._theGCom = aGCom
        }
    }

    public SetResource(source: U3DMobile.PackageSource, pkgName: string, comName: string): void {
        if (this._filled) {
            Log.Error(`can not set package resource repeatedly`)
            return
        }

        if (!pkgName || pkgName.match(/^\s*$/)) {
            Log.Error(`try to set a empty package name`)
            return
        }
        if (!comName || comName.match(/^\s*$/)) {
            Log.Error(`try to set a empty component name`)
            return
        }

        this._filled    = true
        this._pkgSource = source
        this._pkgName   = pkgName
        this._comName   = comName
    }

    public Show(): void {
        if (!this._filled) {
            return
        }

        if (!this._created) {
            //actions:
            if (!this._theGCom) {
                let manager = U3DMobile.PackageManager.instance
                let aGObject = manager.CreateGObject(this._pkgSource, this._pkgName, this._comName)

                this._theGCom = aGObject.asCom
                if (!this._theGCom) {
                    Log.Error(`failed to create '${this._pkgName}/${this._comName}'`)
                    return
                }
            }

            this._window = new FairyGUI.Window()
            this._window.contentPane = this._theGCom

            this._com = new UICom()
            this._com.SetGCom(this._theGCom, this._comName)
            this._com.Bind(this)

            //notification.
            this._created = true
            this.OnCreate()
        }
        if (!this._shown) {
            //notification.
            this._shown = true
            this.OnShow()

            //actions.
            this._window.Show()
        }
    }

    public Hide(): void {
        if (this._shown) {
            //actions.
            this._window.Hide()

            //notification.
            this._shown = false
            this.OnHide()
        }
    }

    public Dispose(): void {
        if (this._shown) {
            //notificaiton.
            this._shown = false
            this.OnHide()
        }
        if (this._created) {
            //actions:
            this._com = null

            this._window.Dispose()
            this._window = null

            this._theGCom = null

            //notification.
            this._created = false
            this.OnDestroy()
        }
    }

    protected OnCreate (): void {}
    protected OnShow   (): void {}
    protected OnHide   (): void {}
    protected OnDestroy(): void {}
}
